README.md
## Barnes - GC Statsd Reporter

A fork of [trashed](https://github.com/basecamp/trashed) focused on Ruby metrics for Heroku.

## Setup

### Rails 5

On Rails 5 (and Rails 3 and 4), add this to your Gemfile:

```
gem "barnes"
```

Then run:

```
$ bundle install
```

### Non-Rails

Add the gem to the Gemfile

```
gem "barnes"
```

Then run:

```
$ bundle install
```

In your puma.rb file:


```ruby
require 'barnes'
```

Then you'll need to start the client with default values:

```ruby
Barnes.start
```



CHANGELOG.md
## 0.0.7

- Report when an app is using Puma

## 0.0.6

- Support Puma max threads, support Puma spawned threads
- Drop Puma backlog metric

## 0.0.5

- Support Puma pool capacity value #15

## 0.0.4

- Support Puma backlog #14

## 0.0.3

## 0.0.2

- Fix requires for non-rails apps (#11)

## 0.0.1

- Fork and Release


test/test_helper.rb
require 'barnes'
require 'minitest/autorun'
require 'pathname'

require 'wait_for_it'
require 'simplecov'
SimpleCov.start

def fixture_path(path = "")
  return Pathname.new(__dir__).join("fixtures").join(path).expand_path
end


require 'socket'
require 'timeout'

def is_port_open?(ip, port)
  begin
    Timeout::timeout(1) do
      begin
        s = TCPServer.new(ip, port)
        s.listen(1024)
        s.close
        return true
      rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH => e
        return false
      end
    end
  rescue Timeout::Error
  end

  return false
end

def next_open_port
  (9292...10000).each do |port|
    if is_port_open?('localhost', port)
      return port
    end
  end
  raise "Could not find a port"
end


test/reporter_test.rb
require 'test_helper'
require 'barnes/reporter'
require 'logger'
require 'stringio'

class ReporterTest < Minitest::Test
  class Statsd
    def initialize(batcher)
      @batcher = batcher
    end
    def batch
      yield @batcher
    end
  end

  def test_report_statsd
    batch = MiniTest::Mock.new

    batch.expect :count, true, [:'Rack.Server.All.GC.allocated_objects', 10, 1.0]
    batch.expect :gauge, true, [:'Rack.Server.All.Time.pct.cpu', 9.1, 1.0]

    statsd = Statsd.new batch

    reporter = Barnes::Reporter.new(statsd: statsd, sample_rate: 1)
    reporter.report_statsd \
                Barnes::COUNTERS => { :'GC.allocated_objects' => 10 }, \
                Barnes::GAUGES => { :'Time.pct.cpu' => 9.1 }

    batch.verify
  end
end


test/integration_test.rb
require 'test_helper'

class IntegrationTest < Minitest::Test
  def test_puma
    Dir.chdir(fixture_path("rack/hello_world")) do |dir|
      WaitForIt.new("bundle exec puma", options) do |spawn|
        spawn.wait('"barnes.gauges":', 1)
        expect_spawn_to_contain(spawn, '"threads.spawned":0')
        expect_spawn_to_contain(spawn, '"using.puma":1')
      end
    end
  end

  def test_no_puma
    Dir.chdir(fixture_path("rack/no_puma")) do |dir|
      WaitForIt.new("bundle exec rackup", options) do |spawn|
        spawn.wait('"barnes.gauges":', 1)
        spawn_should_not_contain(spawn, '"using.puma":1')
      end
    end
  end

  def options
    port    = next_open_port
    options = {}
    options[:env] = {
      "DYNO"         => "web.1",
      "PORT"         => port,
      "BARNES_DEBUG" => "1"
    }
    options[:wait_for] = "Use Ctrl-C to stop"
    options
  end

  def spawn_should_not_contain(spawn, string)
    expect_spawn_to_contain(spawn, string, does_contain: false)
  end

  def expect_spawn_to_contain(spawn, string, does_contain: true)
    assert_equal does_contain, !!spawn.contains?(string), "Expected #{spawn.log.read} to contain #{string.inspect} but it did not"
  end
end


lib/barnes.rb
# Copyright (c) 2017 Salesforce
# Copyright (c) 2009 37signals, LLC
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

module Barnes
  DEFAULT_INTERVAL           = 10
  DEFAULT_AGGREGATION_PERIOD = 60
  DEFAULT_STATSD             = :default
  DEFAULT_PANELS             = []


  # Starts the reporting client
  #
  # Arguments:
  #
  #   - interval: How often, in seconds, to instrument and report
  #   - aggregation_period: The minimal aggregation period in use, in seconds.
  #   - statsd: The statsd reporter. This should be an instance of statsd-ruby
  #   - panels: The instrumentation "panels" in use. See `resource_usage.rb` for
  #     an example panel, which is the default if none are provided.
  def self.start(interval: DEFAULT_INTERVAL, aggregation_period: DEFAULT_AGGREGATION_PERIOD, statsd: DEFAULT_STATSD, panels: DEFAULT_PANELS)
    require 'statsd'
    statsd_client = statsd
    panels        = panels
    sample_rate   = interval.to_f / aggregation_period.to_f

    if statsd_client == :default && ENV["PORT"]
      statsd_client = Statsd.new('127.0.0.1', ENV["PORT"])
    end

    if statsd_client && statsd_client != :default
      reporter = Barnes::Reporter.new(statsd: statsd_client, sample_rate: sample_rate)

      unless panels.length > 0
        panels << Barnes::ResourceUsage.new(sample_rate)
      end

      Periodic.new reporter: reporter, sample_rate: sample_rate, panels: panels
    end
  end
end

require 'barnes/reporter'
require 'barnes/resource_usage'
require 'barnes/periodic'
require 'barnes/railtie' if defined? ::Rails::Railtie


lib/barnes/railtie.rb
# Copyright (c) 2017 Salesforce
# Copyright (c) 2009 37signals, LLC
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

require 'rails/railtie'

module Barnes
  # Automatically configures barnes to run with
  # rails 3, 4, and 5. Configuration can be changed
  # in the application.rb. For example
  #
  #   module YourApp
  #     class Application < Rails::Application
  #     config.barnes[:interval] = 20
  #
  class Railtie < ::Rails::Railtie
    config.barnes = {
      interval:           DEFAULT_INTERVAL,
      aggregation_period: DEFAULT_AGGREGATION_PERIOD,
      statsd:             DEFAULT_STATSD,
      panels:             DEFAULT_PANELS,
    }

    initializer 'barnes' do |app|
      Barnes.start(config.barnes)
    end
  end
end


lib/barnes/version.rb
# Copyright (c) 2017 Salesforce
# Copyright (c) 2009 37signals, LLC
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

module Barnes
  VERSION = "0.0.7"
end


lib/barnes/periodic.rb
# Copyright (c) 2017 Salesforce
# Copyright (c) 2009 37signals, LLC
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

require 'barnes/consts'

module Barnes
  # The periodic class is used to send occasional metrics
  # to a reporting instance of `Barnes::Reporter` at a semi-regular
  # rate.
  class Periodic
    def initialize(reporter:, sample_rate: 1, panels: [])
      @reporter = reporter
      @reporter.sample_rate = sample_rate

      # compute interval based on a 60s reporting phase.
      @interval = sample_rate * 60.0
      @panels = panels

      @thread = Thread.new {
        Thread.current[:barnes_state] = {}

        @panels.each do |panel|
          panel.start! Thread.current[:barnes_state]
        end

        loop do
          begin
            sleep @interval

            # read the current values
            env = {
              STATE    => Thread.current[:barnes_state],
              COUNTERS => {},
              GAUGES   => {}
            }

            @panels.each do |panel|
              panel.instrument! env[STATE], env[COUNTERS], env[GAUGES]
            end
            @reporter.report env
          end
        end
      }
      @thread.abort_on_exception = true
    end

    def stop
      @thread.exit
    end
  end
end


lib/barnes/resource_usage.rb
# Copyright (c) 2017 Salesforce
# Copyright (c) 2009 37signals, LLC
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

require 'barnes/panel'

module Barnes
  class ResourceUsage < Panel
    def initialize(sample_rate)
      super()

      require 'barnes/instruments/puma_instrument'
      backlog_reporter = Barnes::Instruments::PumaInstrument.new

      if backlog_reporter.valid?
        instrument backlog_reporter
      end

      require 'barnes/instruments/stopwatch'
      instrument Barnes::Instruments::Stopwatch.new

      if GC.respond_to? :enable_stats
        require 'barnes/instruments/ree_gc'
        instrument Barnes::Instruments::Ruby18GC.new
      end

      # Ruby 1.9+
      if ObjectSpace.respond_to? :count_objects
        require 'barnes/instruments/object_space_counter'
        instrument Barnes::Instruments::ObjectSpaceCounter.new
      end

      # Ruby 1.9+
      if GC.respond_to?(:stat)
        require 'barnes/instruments/ruby_gc'
        instrument Barnes::Instruments::RubyGC.new(sample_rate)
      end

      # Ruby 2.1+ with https://github.com/tmm1/gctools
      if defined? GC::OOB
        require 'barnes/instruments/gctools_oobgc'
        instrument Barnes::Instruments::GctoolsOobgc.new
      end
    end
  end
end


lib/barnes/instruments/puma_backlog.rb
# frozen_string_literal: true

module Barnes
  module Instruments
    class PumaBacklog
      # This class is responsible for consuming a puma
      # generated stats hash that can come in two "flavors"
      # one is a "single process" server which will look like this:
      #
      #    { "backlog": 0, "running": 0, "pool_capacity": 16 }
      #
      # The other is a multiple cluster server that will look like this:
      #
      #    {"workers"=>2, "phase"=>0, "booted_workers"=>2, "old_workers"=>0, "worker_status"=>[{"pid"=>35020, "index"=>0, "phase"=>0, "booted"=>true, "last_checkin"=>"2018-05-21T19:53:18Z", "last_status"=>{"backlog"=>0, "running"=>5, "pool_capacity"=>5}}, {"pid"=>35021, "index"=>1, "phase"=>0, "booted"=>true, "last_checkin"=>"2018-05-21T19:53:18Z", "last_status"=>{"backlog"=>0, "running"=>5, "pool_capacity"=>5}}]}
      #
      class StatValue
        attr_reader :stats, :key

        def initialize(stats, key)
          @stats   = stats
          @key     = key
          @cluster = stats.key?("worker_status")
        end

        def single?
          !cluster?
        end

        def cluster?
          @cluster
        end

        # For single worker process use value directly
        # for multiple workers use the sum.
        #
        # https://github.com/puma/puma/pull/1532
        def value
          return stats[key] if single?
          first_worker = stats["worker_status"].first
          return nil unless first_worker && first_worker["last_status"].key?(key)

          value = 0
          stats["worker_status"].each do |worker_status|
            value += worker_status["last_status"][key] || 0
          end
          return value
        end
      end

      def initialize(sample_rate=nil)
        @debug = ENV["BARNES_DEBUG_PUMA_STATS"]
      end

      def valid?
        defined?(Puma) &&
          Puma.respond_to?(:stats) &&
          ENV["DYNO"] && ENV["DYNO"].start_with?("web")
      end

      def start!(state)
        require 'multi_json'
      end

      def json_stats
        MultiJson.load(Puma.stats || "{}")

      # Puma loader has not been initialized yet
      rescue NoMethodError => e
        raise e unless e.message =~ /nil/
        raise e unless e.message =~ /stats/
        return {}
      end

      def instrument!(state, counters, gauges)
        stats = json_stats
        return if stats.empty?

        puts "Puma debug stats from barnes: #{stats}" if @debug

        pool_capacity = StatValue.new(stats, "pool_capacity").value
        max_threads   = StatValue.new(stats, "max_threads").value
        spawned       = StatValue.new(stats, "running").value

        gauges[:'pool.capacity']    = pool_capacity if pool_capacity
        gauges[:'threads.max']      = max_threads   if max_threads
        gauges[:'threads.spawned']  = spawned       if spawned
      end
    end
  end
end


lib/barnes/instruments/puma_instrument.rb
# frozen_string_literal: true

module Barnes
  module Instruments
    class PumaInstrument
      # This class is responsible for consuming a puma
      # generated stats hash that can come in two "flavors"
      # one is a "single process" server which will look like this:
      #
      #    { "backlog": 0, "running": 0, "pool_capacity": 16 }
      #
      # The other is a multiple cluster server that will look like this:
      #
      #    {"workers"=>2, "phase"=>0, "booted_workers"=>2, "old_workers"=>0, "worker_status"=>[{"pid"=>35020, "index"=>0, "phase"=>0, "booted"=>true, "last_checkin"=>"2018-05-21T19:53:18Z", "last_status"=>{"backlog"=>0, "running"=>5, "pool_capacity"=>5}}, {"pid"=>35021, "index"=>1, "phase"=>0, "booted"=>true, "last_checkin"=>"2018-05-21T19:53:18Z", "last_status"=>{"backlog"=>0, "running"=>5, "pool_capacity"=>5}}]}
      #
      class StatValue
        attr_reader :stats, :key

        def initialize(stats, key)
          @stats   = stats
          @key     = key
          @cluster = stats.key?("worker_status")
        end

        def single?
          !cluster?
        end

        def cluster?
          @cluster
        end

        # For single worker process use value directly
        # for multiple workers use the sum.
        #
        # https://github.com/puma/puma/pull/1532
        def value
          return stats[key] if single?
          first_worker = stats["worker_status"].first
          return nil unless first_worker && first_worker["last_status"].key?(key)

          value = 0
          stats["worker_status"].each do |worker_status|
            value += worker_status["last_status"][key] || 0
          end
          return value
        end
      end

      def initialize(sample_rate=nil)
        @debug = ENV["BARNES_DEBUG_PUMA_STATS"]
      end

      def valid?
        defined?(Puma) &&
          Puma.respond_to?(:stats) &&
          ENV["DYNO"] && ENV["DYNO"].start_with?("web")
      end

      def start!(state)
        require 'multi_json'
      end

      def json_stats
        MultiJson.load(Puma.stats || "{}")

      # Puma loader has not been initialized yet
      rescue NoMethodError => e
        raise e unless e.message =~ /nil/
        raise e unless e.message =~ /stats/
        return {}
      end

      def instrument!(state, counters, gauges)
        stats = json_stats
        return if stats.empty?

        puts "Puma debug stats from barnes: #{stats}" if @debug

        pool_capacity = StatValue.new(stats, "pool_capacity").value
        max_threads   = StatValue.new(stats, "max_threads").value
        spawned       = StatValue.new(stats, "running").value

        gauges[:'pool.capacity']    = pool_capacity if pool_capacity
        gauges[:'threads.max']      = max_threads   if max_threads
        gauges[:'threads.spawned']  = spawned       if spawned
      end
    end
  end
end
